extensions/wikia/Insights/models/InsightsQuerypageModel.php

```diff
@@ -37,17 +52,132 @@ public function getData() {
+	public function getContent( $params ) {
 		$this->queryPageInstance = $this->getDataProvider();
 
+		/**
+		 * 1. Prepare data of articles - title, last revision, link etc.
+		 */
+		$articlesData = $this->fetchArticlesData();
+
+		/**
+		 * 2. Slice a sorting table to retrieve a page
+		 */
+		$this->prepareParams( $params );
+		if ( !isset( $this->sortingArray ) ) {
+			$this->sortingArray = array_keys( $articlesData );
+		}
+		$ids = array_slice( $this->sortingArray, $this->offset, $this->limit, true );
+
+		/**
+		 * 3. Populate $content array with data for each article id
+		 */
 		$content = [];
-		$res = $this->queryPageInstance->doQuery( $this->offset, $this->limit );
-		if ( $res->numRows() > 0 ) {
-			$content = $this->prepareData( $res );
+		foreach ( $ids as $id ) {
+			$content[] = $articlesData[$id];
 		}
 		return $content;
 	}
 
+	private function prepareParams( $params ) {
+		global $wgMemc;
+
+		if ( isset( $params['sort'] ) && isset( $this->sorting[ $params['sort'] ] ) ) {
+			$this->sortingArray = $wgMemc->get( wfSharedMemcKey( $this->getInsightType(), $params['sort'] ) );
+		}
+
+		if ( isset( $params['offset'] ) ) {
+			$this->offset = intval( $params['offset'] );
+		}
+
+		if ( isset( $params['limit'] ) ) {
+			$this->limit = intval( $params['limit'] );
+		}
+	}
+
+	public function fetchArticlesData() {
+		$cacheKey = wfSharedMemcKey( $this->getInsightType(), 'articlesData' );
+		$this->cacheTtl = WikiaResponse::CACHE_STANDARD * 3;
+		$articlesData = WikiaDataAccess::cache( $cacheKey, $this->cacheTtl, function () {
+			$res = $this->queryPageInstance->doQuery();
+
+			if ( $res->numRows() > 0 ) {
+				$articlesData = $this->prepareData( $res );
+			}
+
+			if ( $this->arePageViewsRequired() ) {
+				$articlesIds = array_keys( $articlesData );
+				$pageViewsData = $this->getPageViewsData( $articlesIds );
+				$articlesData = $this->assignPageViewsData( $articlesData, $pageViewsData );
+			}
+
+			return $articlesData;
+		} );
+
+		return $articlesData;
+	}
+
+	public function getPageViewsData( $articlesIds ) {
+		/**
+		 * Get pv for the last 4 Sundays
+		 */
+		$pvTimes = InsightsHelper::getLastFourTimeIds();
+
+		$pvData = [];
+
+		foreach( $pvTimes as $timeId ) {
+			$pvData[] = DataMartService::getPageViewsForArticles( $articlesIds, $timeId );
+		}
+
+		return $pvData;
+	}
+
+	public function createSortingArray( $sortingArray, $key ) {
+		global $wgMemc;
+
+		arsort( $sortingArray, $this->sorting[ $key ] );
+		$cacheKey = wfSharedMemcKey( $this->getInsightType(), $key );
+
+		$wgMemc->set( $cacheKey, array_keys( $sortingArray ), $this->cacheTtl );
+	}
+
+	public function assignPageViewsData( $articlesData, $pageViewsData ) {
+		$sortingData = [];
+
+		foreach ( $articlesData as $articleId => $data ) {
+			$pv = [
+				intval( $pageViewsData[0][ $articleId ] ),
+				intval( $pageViewsData[1][ $articleId ] ),
+				intval( $pageViewsData[2][ $articleId ] ),
+				intval( $pageViewsData[3][ $articleId ] ),
```

Comment:
There shouldn't be. If a page has 0 views, there is just no entry in the db. In this case evaluating false to 0 with `intval()` is not an unexpected behavior that should be logged.